crowbar_framework/lib/crowbar/lock/shared_non_blocking.rb

```diff
@@ -0,0 +1,139 @@
+    # intervallic runs triggered by the chef-client daemon, to prevent
+    # them from interfering with the orchestration of the proposals
+    # being applied:
+    #
+    #   https://bugzilla.suse.com/show_bug.cgi?id=857375
+    #
+    # It has the same semantics as a shared read/write lock, except
+    # that instead of readers reading, we have threads applying
+    # proposals, and instead of a single writer writing, we have the
+    # chef-client daemon triggering a chef-client run.
+
+    attr_reader :node, :owner, :reason, :executor
+
+    def initialize(options = {})
+      super
+      %w(node owner reason).each do |option|
+        unless options[option.to_sym]
+          raise "#{self.class} can't be instantiated without " + \
+                ":#{option} option (#{options.inspect})"
+        end
+        instance_variable_set "@#{option}", options[option.to_sym]
+      end
+
+      # We manage execution of the command via dependency injection,
+      # so that the test suite can execute the commands locally.
+      default_executor = lambda do |node, cmd, shell|
+        RemoteNode.pipe_ssh_cmd(node, cmd) do |pipe|
+          pipe.write(shell)
+        end
+        unless $?.success?
+          raise "pid #{$?.pid} returned #{$?.exitstatus}"
+        end
+      end
+      @executor = options.fetch :executor, default_executor
+
+      # To implement shared locking semantics, we keep track of all
+      # "readers" via a 'foo.lock.d/' directory which contains one
+      # file per "reader":
+      @readers_dir = path + '.d'
+
+      # In order to provide a lock for the writer, we also ensure that
+      # a 'foo.lock' file exists if and only if the 'foo.lock.d/'
+      # directory contains one or more readers.  Doing this safely
+      # requires a "meta-lock" in order to avoid races when we create
+      # or remove the 'foo.lock' file.  We implement that meta-lock
+      # using flock(1) on another 'foo.lock.meta' file.
+      @meta_lock = path + '.meta'
+    end
+
+    def acquire
+      if locked?
+        logger.warn("#{owner} tried to re-acquire remote lock #{path} on #{node}")
+        return
+      end
+
+      logger.debug("#{owner} acquiring remote lock #{path} on #{node}")
+      shell = <<-EOSHELL.strip
+        mkdir -p "#{@readers_dir}"
+        file="#{@readers_dir}/#{owner}"
+        if [ -e "$file" ]; then
+          echo "ERROR: #{owner} already had lock via $file" >&2
+          exit 1
+        fi
+        echo "#{reason}" > "$file"
+        echo "Last required by #{owner}" > "#{path}"
+      EOSHELL
+      run_with_exclusive_lock(shell)
+      logger.debug("#{owner} acquired remote lock #{path} on #{node}")
+      @locked = true
+      self
+    end
+
+    def release
+      logger.debug("#{owner} releasing remote lock #{path} on #{node}")
+      shell = <<-EOSHELL.strip
+        rm -f #{@readers_dir}/#{owner}
+        if [ -d #{@readers_dir} ] && ! ls #{@readers_dir} | grep -q .; then
```

Comment:
Yes, isn't that reasonably obvious? I could add a comment but then that's an extra line of spam appearing in `production.log` every time a lock is released, since the shell code is logged at DEBUG level ...